core/src/main/java/org/apache/iceberg/avro/ValueReaders.java

```diff
@@ -654,21 +673,29 @@ public S read(Decoder decoder, Object reuse) throws IOException {
       S struct = reuseOrCreate(reuse);
 
       if (decoder instanceof ResolvingDecoder) {
-        // this may not set all of the fields. nulls are set by default.
+        Set<Integer> existingFieldPositionsSet = Sets.newHashSet();
+        // This may not set all of the fields. nulls are set by default.
         for (Schema.Field field : ((ResolvingDecoder) decoder).readFieldOrder()) {
           Object reusedValue = get(struct, field.pos());
           set(struct, field.pos(), readers[field.pos()].read(decoder, reusedValue));
+          existingFieldPositionsSet.add(field.pos());
+        }
+        // Set default values
```

Comment:
Style: missing empty newline between control flow block and the following statement.